core/solver/gcr.cpp

```diff
@@ -0,0 +1,358 @@
+    GKO_SOLVER_STOP_REDUCTION_ARRAYS();
+
+    // Initialization
+    // residual = dense_b
+    // reset stop status
+    exec->run(gcr::make_initialize(dense_b, residual, stop_status.get_data()));
+    // residual = residual - Ax
+    // Note: x is passed in with initial guess
+    this->get_system_matrix()->apply(neg_one_op, dense_x, one_op, residual);
+    this->get_preconditioner()->apply(residual, precon_residual);
+    // A_precon_residual = A*precon_residual
+    this->get_system_matrix()->apply(precon_residual, A_precon_residual);
+
+    // p(:, 1) = precon_residual(:, 1)
+    // Ap(:, 1) = A_precon_residual(:, 1)
+    // final_iter_nums = {0, ..., 0}
+    // apply preconditioner to residual
+    exec->run(gcr::make_restart(precon_residual, A_precon_residual,
+                                krylov_bases_p, mapped_krylov_bases_Ap,
+                                final_iter_nums.get_data()));
+
+    auto stop_criterion = this->get_stop_criterion_factory()->generate(
+        this->get_system_matrix(),
+        std::shared_ptr<const LinOp>(dense_b, [](const LinOp*) {}), dense_x,
+        residual);
+
+    int total_iter = -1;
+    size_type restart_iter = 0;
+
+    /* Memory movement summary for average iteration with krylov_dim d:
+     * (5d+23+6/d)n * values + matrix/preconditioner stroage
+     * 1x SpMV:                       2n * values + storage
+     * 1x Preconditioner:             2n * values + storage
+     * MGS:                   (5d + 19)n = sum k=0 to d-1 of (10k+24)n/d
+     *       1x dots             2(k+1)n in iteration k (0-based)
+     *       1x scals            2(k+1)n in iteration k (0-based)
+     *       2x axpys            6(k+1)n in iteration k (0-based)
+     *       1x norm2                  n
+     *       1x dot                   2n
+     *       1x sq_norm2               n
+     *       1x step 1 (scal, axpys)  6n
+     *       2x copy                  4n
+     * Restart:                   (6/d)n  (every dth iteration)
+     *       3x copy                  6n
+     */
+    while (true) {
+        ++total_iter;
+        // compute residual norm
+        residual->compute_norm2(residual_norm, reduction_tmp);
+        // Log current iteration
+        this->template log<log::Logger::iteration_complete>(
+            this, total_iter, residual, dense_x, residual_norm);
+        // Check stopping criterion
+        if (stop_criterion->update()
+                .num_iterations(total_iter)
+                .residual(residual)
+                .residual_norm(residual_norm)
+                .solution(dense_x)
+                .check(RelativeStoppingId, true, &stop_status, &one_changed)) {
+            break;
+        }
+
+        // If krylov_dim reached, restart with new initial guess
+        if (restart_iter == krylov_dim) {
+            // Restart
+            // p(:, 1) = precon_residual(:, 1)
+            // Ap(:, 1) = A_precon_residual(:, 1)
+            // final_iter_nums = {0, ..., 0}
```

Comment:
I am not sure what this tries to say.

Reply:
Are you asking about the `final_iter_nums`? I copied this part, but I believe this saves the iteration numbers of each RHS, so it has to be reset while doing a restart. Same as GMRES.

Reply:
`final_iter_nums` is the iteration until convergence for the method. `{0,...,0}` shows that it is initialized for multiple right hand side.